crates/iota-grpc-api/src/lib.rs

```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+fn default_grpc_api_address() -> std::net::SocketAddr {
+    use std::net::{IpAddr, Ipv4Addr};
+    std::net::SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 50051)
+}
+
+fn default_checkpoint_broadcast_buffer_size() -> usize {
+    100
+}
+
+pub mod node {
+    tonic::include_proto!("iota.grpc");
+}
+
+use iota_types::storage::RestStateReader;
+use node::{BcsData, node_service_server::NodeService};
+pub mod client;
+use iota_grpc_types::{
+    CertifiedCheckpointSummary as GrpcCertifiedCheckpointSummary,
+    CheckpointData as GrpcCheckpointData,
+};
+use iota_types::{
+    full_checkpoint_content::CheckpointData, messages_checkpoint::CertifiedCheckpointSummary,
+};
+use tracing::{debug, info};
+
+type Receiver<T> = tokio::sync::broadcast::Receiver<T>;
+
+impl BcsData {
+    fn serialize_from<T>(data: &T) -> Result<Self, bcs::Error>
+    where
+        T: Serialize,
+    {
+        let serialized = bcs::to_bytes(data)?;
+        Ok(BcsData { data: serialized })
+    }
+
+    fn deserialize_into<T>(&self) -> Result<T, bcs::Error>
+    where
+        T: for<'de> Deserialize<'de>,
+    {
+        bcs::from_bytes(&self.data)
+    }
+}
+
+pub struct NodeGrpcService {
+    pub state_reader: Arc<dyn RestStateReader>,
+    pub grpc_checkpoint_summary_tx:
+        tokio::sync::broadcast::Sender<Arc<GrpcCertifiedCheckpointSummary>>,
+    pub grpc_checkpoint_data_tx: tokio::sync::broadcast::Sender<Arc<GrpcCheckpointData>>,
+}
+
+impl NodeGrpcService {
+    pub fn new(
+        state_reader: Arc<dyn RestStateReader>,
+        grpc_checkpoint_summary_tx: tokio::sync::broadcast::Sender<
+            Arc<GrpcCertifiedCheckpointSummary>,
+        >,
+        grpc_checkpoint_data_tx: tokio::sync::broadcast::Sender<Arc<GrpcCheckpointData>>,
+    ) -> Self {
+        Self {
+            state_reader,
+            grpc_checkpoint_summary_tx,
+            grpc_checkpoint_data_tx,
+        }
+    }
+}
+
+// Checkpoint stream item.
+// Note, node::Checkpoint may contain either checkpoint data or summary.
+type CheckpointStreamResult = Result<node::Checkpoint, Status>;
+
+// Helper trait for getting checkpoint data and summaries,
+// intended as an abstraction for Arc<dyn RestStateReader>.
+trait CheckpointReader<T>
+where
+    T: Send + Sync + 'static + Serialize,
+    Self: Send + Sync + 'static,
+{
+    fn get_sequence_number(&self, item: &Arc<T>) -> u64;
+    fn get_item(&self, ix: u64) -> Option<Arc<T>>;
+    fn get_latest(&self) -> Option<u64>;
+
+    fn create_checkpoint_response(&self, item: &Arc<T>, is_full: bool) -> CheckpointStreamResult {
+        BcsData::serialize_from(item)
+            .map(|data| node::Checkpoint {
+                sequence_number: self.get_sequence_number(item),
+                bcs_data: Some(data),
+                is_full,
+            })
+            .map_err(|e| Status::internal(format!("BCS serialization error: {e}")))
+    }
+}
+
+#[derive(Clone)]
+struct Reader {
+    state_reader: Arc<dyn RestStateReader>,
```

Comment:
Does this currently add a dependency on the rest api being enabled and the indexes maintained?

Reply:
I ask because of
```
/// Trait used to provide functionality to the REST API service.
///
/// It extends both ObjectStore and ReadStore by adding functionality that may
/// require more detailed underlying databases or indexes to support.
```
but I guess all of those are behind the RestIndexes